### PR TITLE
Remove Bash from Default Foreign Shells

### PR DESCRIPTION
**Pull request by scopatz**
This removes Bash from the default foreign shells in an attempt to reduce the startup time.  This adds some platform specific code for the default `$PATH`, and so it probably requires some testing on environments other than my own. @melund @gforsyth @Carreau @OllieTerrance, @ellocogato, @jivanyatra - your feedback and help would be super appreciated.

Here are some timing comparisons on my desktop:

```
# interactive login shell, master
scopatz@artemis ~ $ time -p xonsh --shell-type=readline -l -i -c "print(42)"
42
real 0.35
user 0.33
sys 0.01

# interactive login shell, this branch
scopatz@artemis ~ $ time -p xonsh --shell-type=readline -l -i -c "print(42)"
42
real 0.28
user 0.27
sys 0.00

-------------

# headless shell, master
scopatz@artemis ~ $ time -p xonsh --shell-type=readline -c "print(42)"
42
real 0.28
user 0.27
sys 0.00

# headless shell, this branch
scopatz@artemis ~ $ time -p xonsh --shell-type=readline -c "print(42)"
42
real 0.28
user 0.26
sys 0.01
```

So as expected it doesn't really effect running scripts, but it knocks about 20% off of the startup time of the command line REPL, getting us ever closer to my goal of 0.1 sec :)

### CODE REVIEW

**Comment by scopatz** [comment]
Ok - tests pass.  Sorry about that!

**Comment by adqm** [comment]
Looks good to me on the two GNU/Linux machines I've tested.  I can try on Windows and Cygwin as well.

I saw similar things: about .1 seconds difference on login/interactive shells, and no difference on headless.

Do we still want to keep the bash completer loaded and enabled by default?  That might get us some more speed...but, of course, we lose the niceties associated with the bash completer.  And then again, loading the cached bash completion functions isn't too slow.

**Comment by scopatz** [comment]
> Do we still want to keep the bash completer loaded and enabled by default? That might get us some more speed...but, of course, we lose the niceties associated with the bash completer. And then again, loading the cached bash completion functions isn't too slow.

I think that if we need to we can move that to another thread or to the first TAB press.  It isn't something that is critical for startup.  But as you say, it isn't think thing getting in our way right now.

**Comment by adqm** [comment]
For me, similar results on Windows and Cygwin (faster by about .1 second on each), and everything still seems to be working in interactive mode.  All looks good to me!  But we should definitely get more eyes on it.

I guess I am a little bit worried about what happens if the Path lookup fails on Windows (or if M$ decided to move where the path is stored in the registry between versions, or something like that).

**Comment by scopatz** [comment]
Yeah, more eyes are definitely needed.  And yeah, maybe we could have a sensible windows fallback.

**Comment by mitnk** [comment]
tested on mac os x.

on master:

```
hugo@mbp: ~$ time -p xonsh --shell-type=readline -l -i -c "print(42)"
42
real         0.42
user         0.33
sys          0.09
hugo@mbp: ~$ time -p xonsh --shell-type=readline -l -i -c "print(42)"
42
real         0.41
user         0.33
sys          0.09
hugo@mbp: ~$ time -p xonsh --shell-type=readline -l -i -c "print(42)"
42
real         0.40
user         0.32
sys          0.08
hugo@mbp: ~$ time -p xonsh --shell-type=readline -c "print(42)"
42
real         0.40
user         0.33
sys          0.07
hugo@mbp: ~$ time -p xonsh --shell-type=readline -c "print(42)"
42
real         0.40
user         0.33
sys          0.08
hugo@mbp: ~$ time -p xonsh --shell-type=readline -c "print(42)"
42
real         0.40
user         0.33
sys          0.07
```

on rmbash

```
hugo@mbp: ~$ time -p xonsh --shell-type=readline -l -i -c "print(42)"
42
real         0.42
user         0.34
sys          0.09
hugo@mbp: ~$ time -p xonsh --shell-type=readline -l -i -c "print(42)"
42
real         0.44
user         0.36
sys          0.10
hugo@mbp: ~$ time -p xonsh --shell-type=readline -l -i -c "print(42)"
42
real         0.41
user         0.33
sys          0.08
hugo@mbp: ~$ time -p xonsh --shell-type=readline -c "print(42)"
42
real         0.40
user         0.33
sys          0.07
hugo@mbp: ~$ time -p xonsh --shell-type=readline -c "print(42)"
42
real         0.39
user         0.32
sys          0.07
hugo@mbp: ~$ time -p xonsh --shell-type=readline -c "print(42)"
42
real         0.39
user         0.32
sys          0.07
```

**Comment by OllieTerrance** [comment]
Cygwin results are in:

```
User@Host ~/xonsh (master) $ time -p xonsh --shell-type=readline -l -i -c "print(42)"
42
real 2.18
user 0.60
sys 0.78

User@Host ~/xonsh (master) $ time -p xonsh --shell-type=readline -c "print(42)"
42
real 0.81
user 0.29
sys 0.48
```

```
User@Host ~/xonsh (rmbash) $ time -p xonsh --shell-type=readline -l -i -c "print(42)"
42
real 1.07
user 0.40
sys 0.37

User@Host ~/xonsh (rmbash) $ time -p xonsh --shell-type=readline -c "print(42)"
42
real 0.98
user 0.37
sys 0.44
```

**Comment by ellocogato** [comment]
My experience on Cygwin reflects that of @OllieTerrance, so I won't clutter the comments with the numbers.  Generally rather slow to begin with, and moderate improvement on the rmbash branch.

File: xonsh/platform.py [hunk 1]
```diff
@@ -212,26 +211,33 @@ def scandir(path):
...
 BASH_COMPLETIONS_DEFAULT = ()
-""" A possibly empty tuple with default paths to Bash completions known for
-    the current platform.
+"""A possibly empty tuple with default paths to Bash completions known for
+the current platform.
 """
 
 if LINUX_DISTRO == 'arch':
     BASH_COMPLETIONS_DEFAULT = (
         '/etc/bash_completion',
```
Comment (gforsyth) [inline xonsh/platform.py]: On Arch this should be `/usr/share/bash-completion/bash_completion`

File: xonsh/platform.py [hunk 2]
```diff
@@ -212,26 +211,33 @@ def scandir(path):
...
         '/usr/share/bash-completion/completions')
+    PATH_DEFAULT = (os.path.expanduser('~/bin'), '/usr/local/sbin',
```
Comment (gforsyth) [inline xonsh/platform.py]: I think for Arch and it's derivatives you can remove `~/bin` from the default `PATH`.  Although it doesn't do any harm being there, but the arch build tools are such that most people will install things into `/usr`

File: xonsh/platform.py (hunk 1 above)
Comment (gforsyth) [inline xonsh/platform.py]: Ahh, not your change, I see.  In any case, this is still incorrect.

**Comment by gforsyth** [comment]
My laptop (running Arch)

```console
gil@bad_cat ~/git/xonsh master $ #MASTER
gil@bad_cat ~/git/xonsh master $ #interactive login
gil@bad_cat ~/git/xonsh master $ time -p xonsh --shell-type=readline -
................................ -no-rc -l -i -c "print(42)"
42
real 0.64
user 0.60
sys 0.03
gil@bad_cat ~/git/xonsh pr/1212 $ #RMBASH
gil@bad_cat ~/git/xonsh pr/1212 $ #interactive login
gil@bad_cat ~/git/xonsh pr/1212 $ time -p xonsh --shell-type=readline 
................................. --no-rc -l -i -c "print(42)"
42
real 0.52
user 0.48
sys 0.03
gil@bad_cat ~/git/xonsh master $ #MASTER
gil@bad_cat ~/git/xonsh master $ #headless
gil@bad_cat ~/git/xonsh master $ time -p xonsh --shell-type=readline -
................................ -no-rc -c "print(42)"
42
real 0.52
user 0.46
sys 0.05
gil@bad_cat ~/git/xonsh pr/1212 $ #RMBASH
gil@bad_cat ~/git/xonsh pr/1212 $ #headless
gil@bad_cat ~/git/xonsh pr/1212 $ time -p xonsh --shell-type=readline 
................................. --no-rc -c "print(42)"
42
real 0.50
user 0.47
sys 0.02
```

Desktop at work (running Arch)
```console
gil@theo ~/git/xonsh master $ time -p xonsh --shell-type=readline --no
............................. -rc -l -i -c "print(42)"
42
real 0.47
user 0.40
sys 0.05
gil@theo ~/git/xonsh master $ git checkout pr/1212
Branch pr/1212 set up to track remote branch pr/1212 from origin.
Switched to a new branch 'pr/1212'
gil@theo ~/git/xonsh pr/1212 $ time -p xonsh --shell-type=readline --n
.............................. o-rc -l -i -c "print(42)"
42
real 0.43
user 0.33
sys 0.02
gil@theo ~/git/xonsh pr/1212 $ git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
gil@theo ~/git/xonsh master $ time -p xonsh --shell-type=readline --no
............................. -rc -c "print(42)"
42
real 0.35
user 0.31
sys 0.02
gil@theo ~/git/xonsh master $ git checkout pr/1212
Switched to branch 'pr/1212'
Your branch is up-to-date with 'origin/pr/1212'.
gil@theo ~/git/xonsh pr/1212 $ time -p xonsh --shell-type=readline --n
.............................. o-rc -c "print(42)"
42
real 0.32
user 0.27
sys 0.02
```

Ran each case a couple of times and the timings are representative.

**Comment by scopatz** [comment]
@OllieTerrance @ellocogato - a whole second!

**Comment by scopatz** [comment]
@gforsyth - thanks for those corrections. Updated.

File: xonsh/platform.py [hunk 3]
```diff
@@ -240,7 +246,12 @@ def scandir(path):
                      'usr\\share\\bash-completion\\completions'),
         os.path.join(GIT_FOR_WINDOWS_PATH,
                      'mingw64\\share\\git\\completion\\git-completion.bash'))
-
+    PATH_DEFAULT = tuple(winreg.QueryValueEx(winreg.OpenKey(
+            winreg.HKEY_LOCAL_MACHINE,
+            r'SYSTEM\CurrentControlSet\Control\Session Manager\Environment'),
+        'Path')[0].split(os.pathsep))
```
Comment (melund) [inline xonsh/platform.py]: I haven't followed all the discussions on gitter, but why is this necessary? Don't we still inherit the environment from `os.envrion` on startup? What is the situation where it would use this default value?

**Comment by melund** [comment]
@scopatz . I don't have the `time` command, but it seems to be around 0.25 sec faster with this PR

File: xonsh/platform.py (hunk 3 above)
Comment (scopatz) [inline xonsh/platform.py]: We do inherit from the calling environment, as you say @melund, so the use case is rare (but important).  Suppose that the calling env doesn't have a $PATH, then we should still be able to run programs by having a reasonable fallback. 

For example, to test this, I was doing something like:

```
xonsh level 0 $ echo $PATH
inherited:path:values
xonsh level 0 $ del $PATH
xonsh level 0 $ xonsh
xonsh level 1 $ echo $PATH
default:path:values
```

**Comment by scopatz** [comment]
@melund, great!

File: xonsh/platform.py (hunk 3 above)
Comment (melund) [inline xonsh/platform.py]: Ahh. I see. I will just try out and see if works.

Comment (melund) [inline xonsh/platform.py]: OK. There is a small bug in `locate_binary()` which prevent this.  I have made a PR for it. #1215

But that aside I don't think Anaconda Python can start on windows unless $PATH exists. Probably Continuum added something to `lib/site.py`. The folder `Library\bin` is prepended to the path on startup. However, it doesn't seem to be an issue with the original python source.

**Comment by scopatz** [comment]
Hey all, Any chance that this could get merged soon. I have spent the past couple of days profiling, not realizing that this wasn't in yet :)

**Comment by gforsyth** [comment]
Wham, bam, thank you @scopatz!

**Comment by scopatz** [comment]
Thanks @gforsyth!
